moveit_core/online_signal_smoothing/src/acceleration_filter.cpp

```diff
@@ -0,0 +1,299 @@
+  move_group_name_ = params.move_group_name;
+
+  // get robot acceleration limits and store in member variables
+  auto joint_model_group = robot_model_->getJointModelGroup(move_group_name_);
+  auto joint_bounds = joint_model_group->getActiveJointModelsBounds();
+  min_acceleration_limits_ = Eigen::VectorXd::Zero(num_joints);
+  max_acceleration_limits_ = Eigen::VectorXd::Zero(num_joints);
+  size_t ind = 0;
+  for (const auto& joint_bound : joint_bounds)
+  {
+    for (const auto& variable_bound : *joint_bound)
+    {
+      if (variable_bound.acceleration_bounded_)
+      {
+        min_acceleration_limits_[ind] = variable_bound.min_acceleration_;
+        max_acceleration_limits_[ind] = variable_bound.max_acceleration_;
+      }
+      else
+      {
+        RCLCPP_ERROR(node_->get_logger(), "The robot must have acceleration joint limits specified for all joint to "
+                                          "use AccelerationLimitedPlugin.");
+        return false;
+      }
+    }
+    ind++;
+  }
+
+  // setup osqp optimization problem
+  Eigen::SparseMatrix<double> P_sparse(1, 1);
+  P_sparse.insert(0, 0) = 1.0;
+  size_t num_constraints = num_joints + 1;
+  A_sparse_ = Eigen::SparseMatrix<double>(num_constraints, 1);
+  for (size_t i = 0; i < num_constraints - 1; ++i)
+  {
+    A_sparse_.insert(i, 0) = 0;
+  }
+  A_sparse_.insert(num_constraints - 1, 0) = 0;
+  osqp_set_default_settings(&settings_);
+  settings_.warm_start = 0;
+  settings_.verbose = 0;
+  data_ = std::make_shared<OSQPDataWrapper>(P_sparse, A_sparse_);
+  data_->q[0] = 0;
+
+  if (osqp_setup(&work_, &data_->data, &settings_) != 0)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+double jointLimitAccelerationScalingFactor(const Eigen::VectorXd& accelerations,
+                                           const moveit::core::JointBoundsVector& joint_bounds)
+{
+  double min_scaling_factor = 1.0;
+
+  // Now get the scaling factor from joint limits.
+  size_t idx = 0;
+  for (const auto& joint_bound : joint_bounds)
+  {
+    for (const auto& variable_bound : *joint_bound)
+    {
+      const auto& target_accel = accelerations(idx);
+      if (variable_bound.acceleration_bounded_ && target_accel != 0.0)
+      {
+        // Find the ratio of clamped acceleration to original acceleration
+        const auto bounded_vel =
+            std::clamp(target_accel, variable_bound.min_acceleration_, variable_bound.max_acceleration_);
+        double joint_scaling_factor = bounded_vel / target_accel;
+        min_scaling_factor = std::min(min_scaling_factor, joint_scaling_factor);
+      }
+      ++idx;
+    }
+  }
+
+  return min_scaling_factor;
+}
+
+bool AccelerationLimitedPlugin::doSmoothing(Eigen::VectorXd& positions, Eigen::VectorXd& velocities,
+                                            Eigen::VectorXd& /* unused */)
+{
+  const size_t num_positions = velocities.size();
+  if (num_positions != num_joints_)
+  {
+    RCLCPP_ERROR_THROTTLE(
```

Comment:
Why does this need to be throttled?

Reply:
I think it is just to prevent the terminal from being spammed by the same error.